torch/testing/_asserts.py

```diff
@@ -370,194 +370,181 @@ def _amend_error_message(exc: E, msg_fmtstr: str) -> E:
     return type(exc)(msg_fmtstr.format(str(exc)))
 
 
+class _TensorPair(NamedTuple):
+    actual: Tensor
+    expected: Tensor
+
+
 _SEQUENCE_MSG_FMTSTR = "The failure occurred at index {} of the sequences."
 _MAPPING_MSG_FMTSTR = "The failure occurred for key '{}' of the mappings."
 
 
-def _check_inputs(
-    actual: Union[Tensor, List[Tensor], Dict[Any, Tensor]],
-    expected: Union[Tensor, List[Tensor], Dict[Any, Tensor]],
-    check_tensors: Callable[[Tensor, Tensor], Optional[Exception]],
+def _check_pair(
+    pair: Union[_TensorPair, List, Dict],
+    check_tensors: Callable[[Any, Any], Optional[Exception]],
 ) -> Optional[Exception]:
-    """Checks inputs.
+    """Checks input pairs.
 
-    :class:`~collections.abc.Sequence`'s and :class:`~collections.abc.Mapping`'s are checked elementwise.
+    :class:`list`'s or :class:`dict`'s are checked elementwise. Checking is performed recursively and thus nested
+    containers are supported.
 
     Args:
-        actual (Union[Tensor, List[Tensor], Dict[Any, Tensor]]): Actual input.
-        expected (Union[Tensor, List[Tensor], Dict[Any, Tensor]]): Expected input.
+        pair (Union[_TensorPair, List, Dict]): Input pair.
         check_tensors (Callable[[Any, Any], Optional[Exception]]): Callable used to check if a tensor pair matches.
             In case it mismatches should return an :class:`Exception` with an expressive error message.
 
     Returns:
         (Optional[Exception]): Return value of :attr:`check_tensors`.
     """
-    if isinstance(actual, collections.abc.Sequence) and isinstance(expected, collections.abc.Sequence):
-        for idx, (actual_t, expected_t) in enumerate(zip(actual, expected)):
-            exc = check_tensors(actual_t, expected_t)
+    if isinstance(pair, list):
```

Comment:
Why the change from sequence to list? What about tuples?

Reply:
We could test for sequences and mappings here, but `_parse_inputs` transforms each sequence to a list and each mapping to a dict.